File: USACO/Winter 2022 Silver/Contest/Bronze/2016USOpen/diamond.py
fin=open("diamond.in")
fout=open("diamond.out","w")
line=list(int(x) for x in fin.readline().strip().split())
n=line[0]
k=line[1]
arr=[]
for i in range(n):
    arr.append(int(fin.readline().strip()))

arr.sort()
res=[]
for i in range(len(arr)):
    for j in range(len(arr)):
        if i!=j:
            if abs(arr[i]-arr[j])<=k:
                res.append(abs(i-j)+1)

fout.write(str(max(res)))
fout.write("\n")

File: USACO/Contest/Silver/2016USOpen/diamond.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>
#include <map>
#include <set>
using namespace std;
vector<int> ptr(int k,vector<int> arr){
    int a=0,b=0;
    map<int,vector<int>> dict1;
    int res=0;
    while (b<arr.size()){
        if(arr[b]-arr[a]>k) a++;
        else{
            dict1[b-a+1]={a,b};
            res=max(res,b-a+1);
            b++;
        }
    }
    return {res,dict1[res][0],dict1[res][1]};
}
int main(){
    ifstream fin("diamond.in");
    ofstream fout("diamond.out");
    int n,k;
    fin>>n>>k;
    vector<int> arr;
    for (int i=0;i!=n;i++){
        int x;
        fin>>x;
        arr.push_back(x);
    }
    sort(arr.begin(),arr.end());
    vector<int> test=ptr(k,arr);
    arr.erase(arr.begin()+test[1],arr.begin()+test[2]);
    fout<<test[0]+ptr(k,arr)[0]<<"\n";
}